Speedup STAC collections endpoint

omad opened the pull request:
As reported in #802, we were experiencing timeouts or extremely slow responses to the `/stac/collections` endpoint. The implementation was doing a lot of extra data transfer and work, by transferring full geometries from PostGIS into Python before simplifying them down to a bbox.

This moves the processing up into PostGIS, and in my testing gives a speedup between 10x and 121x, and potentially more in different environments. I only tested on my laptop.

Running with local PostGIS server, M4 MacBook Pro.

## pytest-benchmark
Loading `/stac/collections` with the data loaded in `test_stac.py`

<img width="1841" height="114" alt="image" src="https://github.com/user-attachments/assets/bdf7f158-baa0-4dcd-92e9-52210cc69f0a" />

Small dataset, but about 10x faster.

## Running datacube-explorer locally
Using a copy of our database with the horrendously complex geometries

Run the server:
```
uv run gunicorn -w 3 --threads 2 -k gthread --config python:cubedash.gunicorn_config 'cubedash:create_app()'
```

Run benchmark with [hey](https://github.com/rakyll/hey)

Before this change:
```
# 2 workers, 4 requests
$ hey -c 2 -n 4 -t 30 http://127.0.0.1:8000/stac/collections

Summary:
  Total:        23.5620 secs
  Slowest:      11.9025 secs
  Fastest:      11.6424 secs
  Average:      11.7767 secs
  Requests/sec: 0.1698
```

After this change:
```
# For fast, defaults to 50 workers and 200 requests
$ hey -t 30 http://127.0.0.1:8000/stac/collections

Summary:
  Total:        9.8161 secs
  Slowest:      2.6334 secs
  Fastest:      0.2301 secs
  Average:      1.8218 secs
  Requests/sec: 20.3746
```

121x faster.

I haven't added any request caching discussed in the issue, that can be done separately.

Fixes #802 

<!-- readthedocs-preview datacube-explorer start -->
----
📚 Documentation preview 📚: https://datacube-explorer--871.org.readthedocs.build/en/871/

<!-- readthedocs-preview datacube-explorer end -->

omad committed:
STAC Collections query performance

omad committed:
Iterating on stac collections query

omad committed:
Speed up (massively) STAC Collections query

omad committed:
Add test for /stac/collections endpoint

omad committed:
Appease mypy types for pystac

pjonsson commented on cubedash/_stac.py:
```diff
@@ -283,11 +283,10 @@ def as_stac_collection(res: CollectionItem) -> pystac.Collection:
-                    else [-180.0, -90.0, 180.0, 90.0]
-                ]
+                # TODO: Find a nicer way to make the typechecker happier
+                # pystac is too specific in wanting a list[float | int]
```
It's not going to help right now, but it's possible pystac would accept a PR that changes the type annotation from list to Sequence.

omad replied:
Yep, I was thinking that's the best solution too.

pjonsson commented:
I started looking at the code but there's a lot of code being changed, and there are still test failures. I will take a look again when the tests are not failing.

omad commented:
> I started looking at the code but there's a lot of code being changed, and there are still test failures. I will take a look again when the tests are not failing.

It's not too much code... 

But there is a massive amount of duplication between the postgis and postgres index modules unfortunately, but using different sqlalchemy table definition styles which makes all the code that's otherwise identical, slightly different. It's unfortunately been added to my mental tidy up list.

pjonsson commented on cubedash/index/postgis/_api.py:
```diff
@@ -376,39 +355,65 @@ def collections_search_query(
                 name="description",
                 description="product description",
-                alchemy_column=collection.c.definition,
+                alchemy_column=ODC_PRODUCT.definition,
```
The corresponding postgres change also fixes the type of the `offset=` parameter.

omad replied:
I think that `SimpleDocField` turns a plain string into a length 1 tuple... but I'm not sure.

This is ugly though, I'll make them match.

pjonsson replied:
I fixed type annotations around that (and a bunch of usages) in datacube-core, I'm quite sure you need the tuple. A bit surprised the type checker wasn't complaining about these lines.

omad replied:
Thanks!

pyright was complaining, but mypy wasn't for some reason.

It's fixed now regardless.

pjonsson commented on cubedash/index/postgis/_api.py:
```diff
@@ -376,39 +355,65 @@ def collections_search_query(
+    ) -> list[Row]:
+        # STAC Collections only hold a bounding box in EPSG:4326, no geometry
+        # Calculate the bounding box on the server, it's far more efficient
+        collection_bbox = func.Box2D(
```
The RHS takes its data from `TimeOverview`, shouldn't this be called `t_o_bbox` (or `overview_bbox`, but that is longer)?

omad replied:
Hmmm. I might add some comments.

The Cubedash Product (which maps to a STAC Collection) doesn't have any bounding box or geometry attached, all the geometries are in the TimeOverview table, grouped by different `period_types`. In this case we use the `period_type=="all"` to get the one that covers all of time.

pjonsson commented on cubedash/index/postgis/_api.py:
```diff
@@ -376,39 +355,65 @@ def collections_search_query(
+                ProductSpatial.time_latest,
+            )
+            .select_from(TimeOverview)
+            .join(ProductSpatial, ProductSpatial.id == TimeOverview.product_ref)
```
Since you have a real database which triggers performance problems at your disposal: I know the order of joins can matter for performance, are these two joins in the right order or should the order be flipped?

omad replied:
The joins have to be in this order due to the relationships of the tables. Also, if it's trivial to switch the order, I'd generally trust the query planner to do the right thing.

```
 Hash Join  (cost=35.24..8906.12 rows=120 width=827) (actual time=0.473..251.316 rows=144 loops=1)
   Hash Cond: ((product.name)::text = (dataset_type.name)::text)
   ->  Hash Join  (cost=12.00..1231.01 rows=138 width=5550) (actual time=0.160..2.085 rows=144 loops=1)
         Hash Cond: (time_overview.product_ref = product.id)
         ->  Index Scan using time_overview_pkey on time_overview  (cost=0.29..1218.93 rows=138 width=5516) (actual time=0.011..1.861 ro
               Index Cond: (period_type = 'all'::overviewperiod)
         ->  Hash  (cost=9.65..9.65 rows=165 width=38) (actual time=0.136..0.136 rows=144 loops=1)
               Buckets: 1024  Batches: 1  Memory Usage: 19kB
               ->  Seq Scan on product  (cost=0.00..9.65 rows=165 width=38) (actual time=0.019..0.104 rows=144 loops=1)
   ->  Hash  (cost=21.44..21.44 rows=144 width=779) (actual time=0.197..0.197 rows=144 loops=1)
         Buckets: 1024  Batches: 1  Memory Usage: 123kB
         ->  Seq Scan on dataset_type  (cost=0.00..21.44 rows=144 width=779) (actual time=0.012..0.125 rows=144 loops=1)
 Planning Time: 0.550 ms
 Execution Time: 251.369 ms
```


```mermaid
erDiagram
    direction LR
    CUBEDASH_TIME_OVERVIEW }o--|| CUBEDASH_PRODUCT : product_id
    "CUBEDASH_PRODUCT" ||--|| "ODC_PRODUCT" : name

```

pjonsson commented on cubedash/index/postgis/_api.py:
```diff
@@ -376,39 +355,65 @@ def collections_search_query(
+        )
+        bbox_array = array(
+            [
+                func.ST_XMin(collection_bbox),
```
Not knowing Postgis, this looks really strange to me, isn't there some function that gets all 4 corners of a 2D box?

omad replied:
It looked really strange to me too, but best I can tell... no, there isn't, this is what everyone does.

The only other way is passing the WKT strings back to Python and parsing the floats out, which, for 4 floats in a box would be totally fine, but doesn't _feel_ totally fine, so I did it this way.

pjonsson commented on cubedash/index/postgis/_api.py:
```diff
@@ -376,39 +355,65 @@ def collections_search_query(
+                func.ST_YMax(collection_bbox),
+            ]
+        )
+        bbox_or_null = case(
```
The RHS uses None, so I think the naming is a bit indirect. But this appears to be a single-use variable that visually looks like it might wrap reasonably well at its use-site, can't the naming issue be avoided by just inlining this variable?

omad replied:
Sure, that'll work.

pjonsson commented on cubedash/index/postgis/_api.py:
```diff
@@ -376,39 +355,65 @@ def collections_search_query(
         if bbox:
             query = query.where(
-                collection.c.footprint_geometry.intersects(func.ST_MakeEnvelope(*bbox))
+                TimeOverview.footprint_geometry.intersects(func.ST_MakeEnvelope(*bbox))
```
There is no test coverage for most of these if-branches, but I'll trust you know what you are doing with these changes.

omad replied:
Yeah, there's a least one bug that I found in these if-branches. The 'time' filter was crashing by trying to apply a python datetime function to a SQLAlchemy Column.

I only added one test in this PR. There's room for more...

In this case, `collection` was a subquery that held the required columns. I've removed the sub-query since it was unnecessary obfuscation, and apply the WHERE clause against the table.

pjonsson commented on cubedash/index/postgis/_api.py:
```diff
@@ -419,7 +424,7 @@ def collections_search_query(
                         title.alchemy_expression.icontains(value),
                         description.alchemy_expression.icontains(value),
-                        collection.c.name.icontains(value),
+                        ODC_PRODUCT.name.icontains(value),
```
Since this is a performance PR: does SQLAlchemy avoid generating extra junk when there is no `q`,  or should there be a:
```python
if q:
    query = query.where(or_(*expressions))
```
instead of line 430?

omad replied:
This whole section is already inside an `if q:` on line 404.

pjonsson replied:
Duh, the danger of doing evening reviews.

pjonsson commented on cubedash/index/postgres/_api.py:
```diff
@@ -422,40 +402,71 @@ def collections_search_query(
+                PRODUCT.c.time_latest,
+            )
+            .select_from(
+                TIME_OVERVIEW.join(
```
I don't know anything about the table layout, are the joins deliberately inside the `select_from()` here but after it in the corresponding postgis code?

omad replied:
They are deliberate, this is pretty much my first time writing SQLAlchemy code. I couldn't get it working with the exact same structure because the types of the Tables are different. In the legacy postgres section they're plain SQLAlchemy Tables. In the newer postgis section they're SQLAlchemy ORM Tables.

Another bug-bear with the postgis section is that it didn't go all the way in defining joins that SQLAlchemy can apply automatically, and it absolutely should. Doing that would probably make this query cleaner.

pjonsson commented on cubedash/index/postgres/_api.py:
```diff
@@ -464,14 +475,14 @@ def collections_search_query(
         query = query.limit(limit).offset(offset)
 
-        with self.index._active_connection() as conn:
+        with self.engine.begin() as conn:
```
This change isn't present in the postgis index.

omad replied:
Whoops, this crept in from the PR I haven't raised yet, that makes this change everywhere.

The `._active_connection()` pattern looks to me like crashes/weird behaviour waiting to happen. As far as I can tell it doesn't do any real context/connection management, and just leaves open connections hanging around in threadlocal variables.

pjonsson replied:
Yeah, keeping that in a separate PR makes it easier to review.

I fully agree about using the active connection looking like something that should be fixed.

pjonsson commented on integration_tests/test_stac.py:
```diff
@@ -707,6 +707,38 @@ def test_stac_collection(stac_client: FlaskClient):
+# supports the full 'Features' specification.
+
+
+@pytest.mark.parametrize("env_name", ("default",), indirect=True)
```
Since this is a performance issue, shouldn't this also be enabled for the postgis backend?

omad replied:
We're not using the PostGIS backend yet, and, there's a LOT of integration tests here that aren't testing against the PostGIS index either. I was afraid I'd fall down another rabbit hole if I started expanding the test coverage.

The table structure and query structure are identical between the two, with the exception of the naming of the String `name` field that is used to connect the cubedash product table with the odc product/agdc dataset_type table.

omad committed:
Address PR comments

- Explain at a high level why we're using the TimeOverview able
- Fix a couple more typing errors,
  - Can't apply datetime_utc to SQLAlchemy Columns
  - SimpleDocField expects `offset` to be a tuple
- Inline a confusingly name variable in the main query
